### PR TITLE
frontend: extend subsidized transfers with missing functionality

### PR DESCRIPTION
**Pull request by GabrielBuragev**
There are still no tests for this file as it requires some effort.

### CODE REVIEW

File: frontend/src/actions/transfers/subsidized-transfer.ts [hunk 1]
```diff
@@ -33,6 +33,22 @@ export class SubsidizedTransfer extends Transfer {
     this.sourceChain.requestManagerAddress = this.originalRequestManagerAddress;
   }
 
+  async withdraw(provider: IEthereumProvider): Promise<void> {
+    if (this.feeSubAddress) {
+      this.sourceChain.requestManagerAddress = this.feeSubAddress;
+    }
+    await super.withdraw(provider);
+    this.sourceChain.requestManagerAddress = this.originalRequestManagerAddress;
+  }
```
Comment (istankovic) [inline frontend/src/actions/transfers/subsidized-transfer.ts:42]: Is there something like Python's `try...finally` in JS?
What happens if `withdraw` throws an exception that is caught in one of the callers?

File: frontend/src/actions/transfers/subsidized-transfer.ts [hunk 2]
```diff
@@ -33,6 +33,22 @@ export class SubsidizedTransfer extends Transfer {
     this.sourceChain.requestManagerAddress = this.originalRequestManagerAddress;
   }
 
+  async withdraw(provider: IEthereumProvider): Promise<void> {
+    if (this.feeSubAddress) {
```
Comment (fredo) [inline frontend/src/actions/transfers/subsidized-transfer.ts:37]: Does this take into account that the fee sub contract is only called upon requests which were created by the feesub?

File: frontend/src/actions/transfers/subsidized-transfer.ts (hunk 1 above)
Comment (GabrielBuragev) [reply frontend/src/actions/transfers/subsidized-transfer.ts:42]: Thats a very good point Ivan. I will wrap it inside a try/catch block.

File: frontend/src/actions/transfers/subsidized-transfer.ts (hunk 2 above)
Comment (GabrielBuragev) [reply frontend/src/actions/transfers/subsidized-transfer.ts:37]: Well, the fact that we are inside a `SubsidizedTransfer` instance is implying that we used a FeeSub contract for that same request.
Whether we should create a normal `Transfer` or a `SubsidizedTransfer` is decided here: https://github.com/beamer-bridge/beamer/blob/main/frontend/src/composables/useTransferRequest.ts#L57-L69

Comment (fredo) [reply frontend/src/actions/transfers/subsidized-transfer.ts:37]: ok makes sense 👍

**Comment by manuelwedler** [comment]
I aligned with Gabriel and due to the issues of overwriting the request manager address, I reimplemented the `SubsidizedTransfer`. This can be reviewed again now.

**Review by GabrielBuragev — COMMENTED** [review]
The way you implemented it looks perfect!

After thinking about the other possible ways to do it i realized we can avoid repeating ourselves (overriding the `call...` functions) by creating an abstraction around the functions inside the `request-manager.ts`. 
This abstraction will contain a `contractAddress` property that will reference to the on-chain contract that is initialized in the constructor. With this, we will only carry one instance of this class inside the `Transfer` or `SubsidizedTransfer` classes and call all the functions via that instance instead of using these functions as separate units.
This way, the `SubsidizedTransfer` will also look way leaner.
Looking forward to hear your thoughts on this. 

I anyways accept your implementation as it is perfect. If you  like my idea too, we can create an issue and optimize it in the future.

**Review by manuelwedler — APPROVED** [review]
Approving as @GabrielBuragev reviewed and he cannot give an approval for his own PR

**Comment by manuelwedler** [comment]
> The way you implemented it looks perfect!
> 
> After thinking about the other possible ways to do it i realized we can avoid repeating ourselves (overriding the `call...` functions) by creating an abstraction around the functions inside the `request-manager.ts`. This abstraction will contain a `contractAddress` property that will reference to the on-chain contract that is initialized in the constructor. With this, we will only carry one instance of this class inside the `Transfer` or `SubsidizedTransfer` classes and call all the functions via that instance instead of using these functions as separate units. This way, the `SubsidizedTransfer` will also look way leaner. If i am not wrong, it is also more memory efficient. Looking forward to hear your thoughts on this.
> 
> I anyways accept your implementation as it is perfect. If you like my idea too, we can create an issue and optimize it in the future.

@GabrielBuragev 
This would mean, that we create a class inside the `request-manager.ts` that would wrap the transaction functions in order to hold state (the `contractAddress`), right? The transaction functions were written in a functional programming style on purpose in the first place. The reasons were simplicity and easier testing.

There are even cases when we want to use different addresses of the request manager for different functions (e.g. fee sub). I think we would run into the same problem as in my comment above regarding the `checkAndUpdateState` function of the `Transfer`.

In my opinion, it is better to have this separation: `Transfer` / `SubsidizedTransfer` holds state, `request-manager` functions (and other contract functions) are pure.

**Comment by manuelwedler** [comment]
Ah ok that makes more sense. But still we would have the issue about the `checkAndUpdateState` function. It should not use the fee sub address. Or how would that be solved if you only have one instance of the new class?

**Comment by GabrielBuragev** [comment]
Ah true, somehow missed that. You are right, sorry for the confusion.

**Pull request merged**